telegram/utils/typing.py

```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python
+#
+# A library that provides a Python interface to the Telegram Bot API
+# Copyright (C) 2015-2020
+# Leandro Toledo de Souza <devs@python-telegram-bot.org>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+"""This module contains custom typing aliases."""
+from typing import Union, Any, Dict, TYPE_CHECKING, IO, Tuple, Optional
+
+if TYPE_CHECKING:
+    from telegram import InputFile, Update
+
+FileLike = Union[IO, 'InputFile']
+"""Either an open file handler or in :class:`telegram.InputFile`."""
```

Comment:
It did not look at this PR in depth but was curious what was in this file. If this just contains types I would prefer to call this while file `types.py` instead of `typing.py` as this is not contain any other type related stuff like `cast` etc. Furthermore `InputFile` seems to be casted to `IO` a lot. I did not look into the exact use cases but maybe it would be a good alternative to make `InputFile` inherit from `IO` and implement the same interface?

Reply:
Renaming sounds reasonable. About casting `InputFile`: You probably mean those snippets in the `send_<media_type>` methods?

```python
        elif InputFile.is_file(video_note):
            video_note = cast(IO, video_note)
            video_note = InputFile(video_note)
```
`InputFile.is_file` just checks, if the input has a `read()` methods so that `InputFile` can handle the file on init. But it's hard to explain to mypy that this is valid, so I casted.

Reply:
Yes I meant those snippets. Hm that sound tricky - I just tested and `isinstance(open("somefile"), typing.IO)` weirdly returns `False` (though checking with `io.IOBase` yields `True`). `Protocols` would be perfect though they only exist in Python 3.8 and later. Can't think of a better solution right now